### PR TITLE
Allow custom rules

### PR DESCRIPTION
**Pull request by kapilreddy**
- Modify run and external-run to accept custom rules
- Create unifier/prep'ed rules inside check-expr and check-reader. This
  makes passing custom rules simpler

### CODE REVIEW

File: src/kibit/driver.clj [hunk 1]
```diff
@@ -30,20 +31,22 @@
...
-    (mapcat (fn [file] (try (check-file file :reporter (name-to-reporter (:reporter options)
-                                                                         cli-reporter))
-                            (catch Exception e
-                              (println "Check failed -- skipping rest of file")
-                              (println (.getMessage e)))))
+    (mapcat (fn [file] (try (check-file file
+                                       :reporter (name-to-reporter (:reporter options)
+                                                                   cli-reporter)
+                                       :rules (or rules all-rules))
+                           (catch Exception e
```
Comment (danielcompton) [inline src/kibit/driver.clj]: Can you make indentation match the previous usage here?

File: src/kibit/driver.clj [hunk 2]
```diff
@@ -30,20 +31,22 @@
...
+(defn run [source-paths rules & args]
   (let [[options file-args usage-text] (apply (partial cli args) cli-specs)
         source-files (mapcat #(-> % io/file find-clojure-sources-in-dir)
                              (if (empty? file-args) source-paths file-args))]
-    (mapcat (fn [file] (try (check-file file :reporter (name-to-reporter (:reporter options)
-                                                                         cli-reporter))
-                            (catch Exception e
-                              (println "Check failed -- skipping rest of file")
-                              (println (.getMessage e)))))
+    (mapcat (fn [file] (try (check-file file
```
Comment (danielcompton) [inline src/kibit/driver.clj]: Can you keep this on the same line?

Comment (kapilreddy) [inline src/kibit/driver.clj]: You mean something like this ?
```
(check-file file :reporter (name-to-reporter (:reporter options)
                                             cli-reporter)
            :rules (or rules all-rules))
```

File: src/kibit/driver.clj (hunk 1 above)
Comment (kapilreddy) [inline src/kibit/driver.clj]: I am not sure what you mean by that. A code block would help.

File: src/kibit/driver.clj (hunk 2 above)
Comment (danielcompton) [inline src/kibit/driver.clj]: Just keeping the indentation and line wrapping the same as it was before. But looking at this more closely I see why you changed the wrapping.

File: src/kibit/driver.clj (hunk 1 above)
Comment (danielcompton) [inline src/kibit/driver.clj]: The previous catch block was indented inline with the start of `(check-file`. So:

```
(try (check-file ...)
    (catch Exception e ...))

**Comment by kapilreddy** [comment]
@danielcompton  I think all the things have been fixed according to your comments. Would it be possible to get this change merged ?

**Comment by danielcompton** [comment]
Hey sorry I haven't had the chance to sit down and look at this properly yet, my weekend was full. I hope to get to this in the next week or so. Sorry for the delay.

**Comment by kapilreddy** [comment]
Hey Daniel, Take your time.

**Comment by danielcompton** [comment]
I've merged this. Thanks!

**Comment by marksto** [comment]
Hi @danielcompton! This one still isn't merged and it looks like an accident. Are there any plans to finally merge this one together with the #141 (a #140-related PR with the documentation update)?

**Comment by port19x** [comment]
I'm not convinced this is a good idea.
If you're competent enough and familiar enough with kibit to write your own custom rule, surely it wouldn't be much of a stretch to assume you could fork the project.
That way you could also open a PR and contribute back to the project more easily.
It would be a shame if through custom rules we have several people implement the same good custom rule that should really be upstream

**Comment by port19x** [comment]
> Counterpoint: using kibit on a specialized or closed-source codebase (at a job, in a personal project, etc) will have opportunities for helpful custom rules that 1) aren't generally useful and/or 2) can't even be shared due to privacy/copyright concerns.
> 
> For example, I want to migrate my codebase from using two custom functions to a single unified custom function. There's no need for kibit in general to have a `(func2 (func1 ?arg1 ?arg2)) (unified-func ?arg1 ?arg2))` rule.

Valid point!
I totally forgot that you might want to have rules for functions outside of clojure.core, especially in larger codebases

**Comment by NoahTheDuke** [comment]
Closing this PR in light of kibit entering maintenance status. (See [this comment](https://github.com/clj-commons/kibit/issues/250#issuecomment-2115570734) for further details.)

**Pull request closed without merging**